debezium-core/src/main/java/io/debezium/relational/history/TableChanges.java

```diff
@@ -0,0 +1,267 @@
+
+        public TableChange(TableChangeType type, Table table) {
+            this.type = type;
+            this.table = table;
+            this.id = table.id();
+        }
+
+        public static TableChange fromDocument(Document document) {
+            TableChangeType type = TableChangeType.valueOf(document.getString("type"));
+            TableId id = TableId.parse(document.getString("id"));
+            Table table = null;
+
+            if (type == TableChangeType.CREATE) {
+                table = fromDocument(id, document.getDocument("table"));
+            }
+
+            return new TableChange(type, table);
+        }
+
+        public TableChangeType getType() {
+            return type;
+        }
+
+        public TableId getId() {
+            return id;
+        }
+
+        public Table getTable() {
+            return table;
+        }
+
+        public Document toDocument() {
+            Document document = Document.create();
+
+            document.setString("type", type.name());
+            document.setString("id", id.toString());
+            document.setDocument("table", toDocument(table));
+            return document;
+        }
+
+        @Override
+        public int hashCode() {
+            final int prime = 31;
+            int result = 1;
+            result = prime * result + id.hashCode();
+            result = prime * result + ((table == null) ? 0 : table.hashCode());
+            result = prime * result + type.hashCode();
+            return result;
+        }
+
+        @Override
+        public boolean equals(Object obj) {
+            if (this == obj)
+                return true;
+            if (obj == null)
+                return false;
+            if (getClass() != obj.getClass())
+                return false;
+            TableChange other = (TableChange) obj;
+            if (!id.equals(other.id))
+                return false;
+            if (table == null) {
+                if (other.table != null)
+                    return false;
+            } else if (!table.equals(other.table))
+                return false;
+            if (type != other.type)
+                return false;
+            return true;
+        }
+
+        @Override
+        public String toString() {
+            return "TableChange [type=" + type + ", id=" + id + ", table=" + table + "]";
+        }
+
+        private Document toDocument(Table table) {
+            Document document = Document.create();
+
+            document.set("defaultCharsetName", table.defaultCharsetName());
+            document.set("primaryKeyColumnNames", Array.create(table.primaryKeyColumnNames()));
+
+            List<Document> columns = table.columns()
+                .stream()
+                .map(this::toDocument)
+                .collect(Collectors.toList());
+
+            document.setArray("columns", Array.create(columns));
+
+            return document;
+        }
+
+        private Document toDocument(Column column) {
```

Comment:
Would not it be better to have an interface like `Documentable` that would be implemented by all classes (value types) that would be able to convert themselves into JSON? The centralization of serialization has its merit but keeping the code for serialization near to the data too - if a new field will be added the code is collocated.
I have no strong preference in this regard just food for thought.

Reply:
Yeah, I was swinging forth and back, too. I think I prefer the separation if different aspects, so to not overload the core classes with too many additional functionality such as serialization.